lightning/src/offers/async_receive_offer_cache.rs

```diff
@@ -0,0 +1,480 @@
+	///
+	/// [`OffersMessageFlow`]: crate::offers::flow::OffersMessageFlow
+	pub fn new() -> Self {
+		Self {
+			offers: Vec::new(),
+			offer_paths_request_attempts: 0,
+			last_offer_paths_request_timestamp: Duration::from_secs(0),
+			paths_to_static_invoice_server: Vec::new(),
+		}
+	}
+
+	pub(super) fn paths_to_static_invoice_server(&self) -> Vec<BlindedMessagePath> {
+		self.paths_to_static_invoice_server.clone()
+	}
+
+	/// Sets the [`BlindedMessagePath`]s that we will use as an async recipient to interactively build
+	/// [`Offer`]s with a static invoice server, so the server can serve [`StaticInvoice`]s to payers
+	/// on our behalf when we're offline.
+	///
+	/// [`StaticInvoice`]: crate::offers::static_invoice::StaticInvoice
+	#[cfg(async_payments)]
+	pub fn set_paths_to_static_invoice_server(
+		&mut self, paths_to_static_invoice_server: Vec<BlindedMessagePath>,
+	) -> Result<(), ()> {
+		if paths_to_static_invoice_server.is_empty() {
+			return Err(());
+		}
+
+		self.paths_to_static_invoice_server = paths_to_static_invoice_server;
+		if self.offers.is_empty() {
+			// See `AsyncReceiveOfferCache::offers`.
+			self.offers = vec![None; MAX_CACHED_OFFERS_TARGET];
+		}
+		Ok(())
+	}
+}
+
+// The target number of offers we want to have cached at any given time, to mitigate too much
+// reuse of the same offer while also limiting the amount of space our offers take up on the
+// server's end.
+#[cfg(async_payments)]
+const MAX_CACHED_OFFERS_TARGET: usize = 10;
+
+// The max number of times we'll attempt to request offer paths per timer tick.
+#[cfg(async_payments)]
+const MAX_UPDATE_ATTEMPTS: u8 = 3;
+
+// If we have an offer that is replaceable and its invoice was confirmed as persisted more than 2
+// hours ago, we can go ahead and refresh it because we always want to have the freshest offer
+// possible when a user goes to retrieve a cached offer.
+//
+// We avoid replacing unused offers too quickly -- this prevents the case where we send multiple
+// invoices from different offers competing for the same slot to the server, messages are received
+// delayed or out-of-order, and we end up providing an offer to the user that the server just
+// deleted and replaced.
+#[cfg(async_payments)]
+const OFFER_REFRESH_THRESHOLD: Duration = Duration::from_secs(2 * 60 * 60);
+
+// Require offer paths that we receive to last at least 3 months.
+#[cfg(async_payments)]
+const MIN_OFFER_PATHS_RELATIVE_EXPIRY_SECS: u64 = 3 * 30 * 24 * 60 * 60;
+
+#[cfg(async_payments)]
+impl AsyncReceiveOfferCache {
+	/// Retrieve a cached [`Offer`] for receiving async payments as an often-offline recipient, as
+	/// well as returning a bool indicating whether the cache needs to be re-persisted.
+	///
+	// We need to re-persist the cache if a fresh offer was just marked as used to ensure we continue
+	// to keep this offer's invoice updated and don't replace it with the server.
+	pub fn get_async_receive_offer(
```

Comment:
Question for reviewers: should we refuse to return offers that don't have at least a week until they expire, or some behavior like that? I'm also wondering if we should always return the offer with the expiry that is the furthest out, regardless of whether it's a used offer or not.

Reply:
Related to what we discussed offline: how is the user going to know that the QR code that they got isn't valid forever? An expiring QR code seems undesirable for a user.

Random association: Is BIP 353 able to do something here somehow? (just writing this down, not familiar with it).

Reply:
The offer expiry is available in the offer, but we may want some behavior in LDK Node where we return the offer expiry as a separate return value to make it more obvious when it expires. To me it seems likely that most offer paths returned by an LSP will be never-expiring, since that's the best UX, but it's hard to be sure right now. 

BIP 353 may indeed help here because IIUC we're putting an offer in a DNS record, so it may be easier to rotate the offer that's in a DNS record than update an offer that's hardcoded into a website.